src/engine/controls/bpmcontrol.cpp

```diff
@@ -274,44 +354,68 @@ void BpmControl::adjustBeatsBpm(double deltaBpm) {
         return;
     }
 
-    const mixxx::Bpm bpm = pBeats->getBpmInRange(
-            mixxx::audio::kStartFramePos, frameInfo().trackEndPosition);
-    // FIXME: calling bpm.value() without checking bpm.isValid()
-    const auto centerBpm = mixxx::Bpm(math_max(kBpmAdjustMin, bpm.value() + deltaBpm));
-    mixxx::Bpm adjustedBpm = BeatUtils::roundBpmWithinRange(
-            centerBpm - kBpmAdjustStep / 2, centerBpm, centerBpm + kBpmAdjustStep / 2);
-    const auto newBeats = pBeats->trySetBpm(adjustedBpm);
-    if (!newBeats) {
-        return;
+    activateActionRepeater([this, v]() {
+        slotAdjustBeatsFaster(v);
+    });
+
+    const auto adjustedBeats =
+            pBeats->tryAdjustTempo(frameInfo().currentPosition,
+                    v > 1 ? mixxx::Beats::TempoAdjustment::MuchFaster
+                          : mixxx::Beats::TempoAdjustment::Faster);
+    if (adjustedBeats) {
+        pTrack->trySetBeats(*adjustedBeats);
     }
-    pTrack->trySetBeats(*newBeats);
 }
 
-void BpmControl::slotAdjustBeatsFaster(double v) {
+void BpmControl::slotAdjustBeatsSlower(double v) {
     if (v <= 0) {
+        clearActionRepeater();
         return;
     }
-    adjustBeatsBpm(kBpmAdjustStep);
-}
 
-void BpmControl::slotAdjustBeatsSlower(double v) {
-    if (v <= 0) {
+    const TrackPointer pTrack = getEngineBuffer()->getLoadedTrack();
+    if (!pTrack) {
         return;
     }
-    adjustBeatsBpm(-kBpmAdjustStep);
+    const mixxx::BeatsPointer pBeats = pTrack->getBeats();
+    if (!pBeats) {
+        return;
+    }
+
+    activateActionRepeater([this, v]() {
+        slotAdjustBeatsSlower(v);
+    });
+
+    const auto adjustedBeats =
+            pBeats->tryAdjustTempo(frameInfo().currentPosition,
+                    v > 1 ? mixxx::Beats::TempoAdjustment::MuchSlower
+                          : mixxx::Beats::TempoAdjustment::Slower);
+    if (adjustedBeats) {
+        pTrack->trySetBeats(*adjustedBeats);
+    }
 }
 
 void BpmControl::slotTranslateBeatsEarlier(double v) {
     if (v <= 0) {
+        clearActionRepeater();
         return;
     }
+
+    activateActionRepeater([this]() {
+        slotTranslateBeatsEarlier(1);
+    });
     slotTranslateBeatsMove(-1);
```

Comment:
> So we may either change the tempo of the previous region, or move them as well.

I strongly disagree; Beatgrids are meant to be contained. If you create a new one, it makes sense to inherit the previous one's BPM, but if you adjust the BPM of it, it should not affect any other adjacent grids. When it comes to BPM.

When it comes to global BPM adjustment, this is an existing limitation of non-const tempo AFAIK, so we may consider a new feature, but clearly out of scope for this PR.

> In general I think these padding beats are a pain.

I don't think padding bar are necessary any more with the beatgrid reword introduced by this PR

Reply:
Clarification on this is a relase blocker. 

> Beatgrids are meant to be contained.  

Confirmed. This means to me that the region before the beat change marker shall not change. 
The issue is here that it does change the last bar of the previous region, introduceing a highly undesired padding beat by default.
 
To find a suitable solution, we need to consider the use case. 

* This can be used if the beat marker is on the "boom" and not on the "stchack"

It is easy possible in case we have only one bea like before. 

In case the user has adjusted the tempo in some regions, it is likely that the whole track is still affected. So it still should be applied to the whole track. 

What other use cases we have?

Reply:
Since you define beatgrid by their start and not their end, it made sense to me to keep the start as immutable. This means that if you shift a beatgrid, the start will move (and thus change the end of the previous grid), but the end will be inferred from the next grid.

Reply:
I don't think we have a use case for this. This will only mess up the beat grid. 

These padding beats don't have a counterpart in the music theory, are undesirable in the sync case and Hart to fix once created. Let's avoid them unless the user explicit wants them.

Or do I miss a case?